File: src/core/stt/SttBackend.ts
import type { SttEngineKind } from '../types';

export interface SttTranscription {
  text: string;
  /** Language the decoder actually reported, when it is multilingual. */
  detectedLanguage?: string;
  /**
   * Text exactly as the engine returned it, before repairScript() ran.
   * Diagnostic-only: additive field for accuracy investigation, not consumed
   * by any production code path.
   */
  rawText?: string;
}

/**
 * A speech recogniser bound to one language.
 *
 * `load` is separated from construction because model loading is slow (hundreds
 * of ms to seconds) and the controller wants to show an INITIALIZING state
 * around it.
 */
export interface SttBackend {
  readonly kind: SttEngineKind;
  /** True once a decoder is resident and `transcribe` can be called. */
  readonly isReady: boolean;
  /** Load (or swap to) the decoder for `languageCode`. */
  load(languageCode: string): Promise<void>;
  /** Decode one utterance of 16 kHz mono audio. */
  transcribe(
    samples: Float32Array,
    languageCode: string
  ): Promise<SttTranscription>;
  dispose(): Promise<void>;
}

File: android-native/app/src/main/java/com/itantra/app/stt/SttBackend.kt
package com.itantra.app.stt

/** Direct port of SttEngineKind in src/core/types.ts. Which concrete STT implementation is live. */
enum class SttEngineKind {
    SHERPA_ONNX,
    SIMULATED,
    NONE,
}

/** Direct port of SttTranscription in src/core/stt/SttBackend.ts. */
data class SttTranscription(
    val text: String,
    /** Language the decoder actually reported, when it is multilingual. */
    val detectedLanguage: String? = null,
    /**
     * Text exactly as the engine returned it, before repairScript() ran.
     * Diagnostic-only, same as the source.
     */
    val rawText: String? = null,
)

/**
 * Direct port of the SttBackend interface in src/core/stt/SttBackend.ts.
 *
 * A speech recogniser bound to one language. load() is separated from
 * construction because model loading is slow (hundreds of ms to
 * seconds).
 */
interface SttBackend {
    val kind: SttEngineKind

    /** True once a decoder is resident and transcribe() can be called. */
    val isReady: Boolean

    /** Load (or swap to) the decoder for [languageCode]. */
    fun load(languageCode: String)

    /** Decode one utterance of 16 kHz mono audio. */
    fun transcribe(samples: FloatArray, languageCode: String): SttTranscription

    fun dispose()
}
